internal/app/repository/SQLStorage.go

```diff
@@ -0,0 +1,138 @@
+package repository
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/jackc/pgerrcode"
+	"github.com/jackc/pgx/v5/pgconn"
+	_ "github.com/jackc/pgx/v5/stdlib"
+)
+
+var (
+	ErrURLAlreadyExists = errors.New("url already exists")
+)
+
+func newErrURLAlreadyExists() error {
+	return ErrURLAlreadyExists
+}
+
+type SQLStorage struct {
+	DB   *sql.DB
+	cnfg string
+}
+
+func NewDB(cnfg string) *SQLStorage {
+	db, err := sql.Open("pgx", cnfg)
+	if err != nil {
+		fmt.Printf("sql.Open error: %v\n", err)
+		return nil
+	}
+
+	if err := db.Ping(); err != nil {
+		fmt.Printf("db.Ping error: %v\n", err)
+		return nil
+	}
+
+	storage := &SQLStorage{
+		DB:   db,
+		cnfg: cnfg,
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	if err := storage.createTableIfNotExists(ctx); err != nil {
+		fmt.Printf("createTableIfNotExists error: %v\n", err)
+		return nil
+	}
+
+	return storage
+}
+
+func (d *SQLStorage) createTableIfNotExists(ctx context.Context) error {
+	_, err := d.DB.ExecContext(ctx, `
+        CREATE TABLE IF NOT EXISTS urls (
+           full_url TEXT UNIQUE,
+           short_url TEXT
+        );
+    `)
+	if err != nil {
+		return fmt.Errorf("error creating table: %w", err)
+	}
+	return nil
+}
+
+func (d *SQLStorage) Close() {
+	if d.DB != nil {
+		d.DB.Close()
+	}
+}
+
+func (d *SQLStorage) SaveURL(u *URL) (*URL, error) {
```

Comment:
```suggestion
func (d *SQLStorage) SaveURL(ctx context.Context, u *URL) (*URL, error) {
```
Контекст при этом берется от запроса (req.Context()), а не из воздуха. И так во всех функциях интерфейса. Для других стораджей можно сделать заглушку `_ context.Context`.